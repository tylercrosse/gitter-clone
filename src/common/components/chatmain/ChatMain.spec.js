import React        from 'react';
import { shallow, mount }  from 'enzyme';
import toJson from 'enzyme-to-json';
import { ChatMain,
  makeMapStateToProps } from './ChatMain';

const setup = (propOverrides) => {
  const props = Object.assign({
    messages: [],
    usersTyping: [],
    user: {
      loggedIn: true
    },
    routeParams: {
      convo: 'chat'
    },
    convoName: 'chat',
    isFetching: false,
    modalIsOpen: {signIn: false},
    error: {
      messages: false
    },
    fetchMessages: jest.fn(),
    addMessage: jest.fn(),
    signIn: jest.fn(),
    signOut: jest.fn(),
    addTypingUser: jest.fn(),
    removeTypingUser: jest.fn(),
    openSignInModal: jest.fn(),
    closeModal: jest.fn()
  }, propOverrides);

  const component = <ChatMain {...props} />;

  return {
    props,
    component
  };
};

describe('<ChatMain />', () => {
  it('should render correctly', () => {
    const { component } = setup();
    const wrapper = shallow(component);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render correctly when fetching', () => {
    const { props } = setup({isFetching: true});
    const wrapper = shallow(<ChatMain {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('should render correctly when 404 error', () => {
    const errorOverride = {
      error: {
        status: 404
      }
    };
    const { props } = setup(errorOverride);
    const wrapper = shallow(<ChatMain {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should call fetchData() at lifecycle changes', () => {
    const { props, component } = setup();
    const newProps = setup({convoName: 'react'}).props;
    const wrapper = mount(component);
    // fetchMessages called once on mount
    expect(props.fetchMessages).toHaveBeenCalledTimes(1);
    // fetchMessages not called, no change in props
    wrapper.instance().componentDidUpdate(props);
    expect(props.fetchMessages).toHaveBeenCalledTimes(1);
    // fetchMessages called again on update
    wrapper.instance().componentDidUpdate(newProps);
    expect(props.fetchMessages).toHaveBeenCalledTimes(2);
  });

  it('should receive the correct props from state', () => {
    const { props } = setup();
    let state = {
      convos: {},
      ui: {isFetching: false},
      error: {messages: false}
    };
    const mapStateToProps = makeMapStateToProps();
    expect(mapStateToProps(state, props))
      .toMatchSnapshot();
    state = {
      convos: {chat: []},
      ui: {isFetching: false},
      error: {messages: false}
    };
    expect(mapStateToProps(state, props))
      .toMatchSnapshot();
  });
});
