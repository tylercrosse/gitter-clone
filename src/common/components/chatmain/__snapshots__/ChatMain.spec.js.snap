// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChatMain /> should receive the correct props from state 1`] = `
Object {
  "convoName": "chat",
  "convos": Object {},
  "error": false,
  "isFetching": false,
  "messages": Array [],
  "modalIsOpen": undefined,
  "user": undefined,
  "usersTyping": undefined,
}
`;

exports[`<ChatMain /> should receive the correct props from state 2`] = `
Object {
  "convoName": "chat",
  "convos": Object {
    "chat": Array [],
  },
  "error": false,
  "isFetching": false,
  "messages": Array [],
  "modalIsOpen": undefined,
  "user": undefined,
  "usersTyping": undefined,
}
`;

exports[`<ChatMain /> should render correctly 1`] = `
<main
  className="chat-header-wrapper"
>
  <Component
    pathname="chat"
    signOut={[Function]}
    user={
      Object {
        "loggedIn": true,
      }
    }
  />
  <div
    className="chat-and-toolbar-wrapper"
  >
    <Component
      messages={Array []}
      user={
        Object {
          "loggedIn": true,
        }
      }
    />
    <div
      className="chat-wrapper"
    >
      <ChatContent
        containerId="chat-content"
        messages={Array []}
      />
      <ChatInput
        addMessage={[Function]}
        addTypingUser={[Function]}
        closeModal={[Function]}
        convoName="chat"
        error={
          Object {
            "messages": false,
          }
        }
        fetchMessages={[Function]}
        isFetching={false}
        messages={Array []}
        modalIsOpen={
          Object {
            "signIn": false,
          }
        }
        onMessageSubmit={[Function]}
        openSignInModal={[Function]}
        removeTypingUser={[Function]}
        routeParams={
          Object {
            "convo": "chat",
          }
        }
        signIn={[Function]}
        signOut={[Function]}
        user={
          Object {
            "loggedIn": true,
          }
        }
        usersTyping={Array []}
      />
    </div>
  </div>
  <SignInModal
    modalIsOpen={false}
    onFormSubmit={[Function]}
    onRequestClose={[Function]}
    user={
      Object {
        "loggedIn": true,
      }
    }
  />
</main>
`;

exports[`<ChatMain /> should render correctly when 404 error 1`] = `<Generic404 />`;

exports[`<ChatMain /> should render correctly when fetching 1`] = `
<main
  className="chat-header-wrapper"
>
  <div
    className="loading-container"
  >
    <div
      className="loading-spinner"
    />
  </div>
  <Component
    pathname="chat"
    signOut={[Function]}
    user={
      Object {
        "loggedIn": true,
      }
    }
  />
  <div
    className="chat-and-toolbar-wrapper"
  >
    <Component
      messages={Array []}
      user={
        Object {
          "loggedIn": true,
        }
      }
    />
    <div
      className="chat-wrapper"
    >
      <ChatContent
        containerId="chat-content"
        messages={Array []}
      />
      <ChatInput
        addMessage={[Function]}
        addTypingUser={[Function]}
        closeModal={[Function]}
        convoName="chat"
        error={
          Object {
            "messages": false,
          }
        }
        fetchMessages={[Function]}
        isFetching={true}
        messages={Array []}
        modalIsOpen={
          Object {
            "signIn": false,
          }
        }
        onMessageSubmit={[Function]}
        openSignInModal={[Function]}
        removeTypingUser={[Function]}
        routeParams={
          Object {
            "convo": "chat",
          }
        }
        signIn={[Function]}
        signOut={[Function]}
        user={
          Object {
            "loggedIn": true,
          }
        }
        usersTyping={Array []}
      />
    </div>
  </div>
  <SignInModal
    modalIsOpen={false}
    onFormSubmit={[Function]}
    onRequestClose={[Function]}
    user={
      Object {
        "loggedIn": true,
      }
    }
  />
</main>
`;
